### PR TITLE
feat!: Total confidence

### PR DESCRIPTION
**Pull request by vahidlazio**

### CODE REVIEW

File: confidence/confidence.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	resolveFlag(ctx context.Context, flag string, defaultValue interface{},
+		evalCtx map[string]interface{}, expectedKind reflect.Kind) InterfaceResolutionDetail
+}
+
+type ContextProvider interface {
+	GetContext() map[string]interface{}
+}
+
+var (
+	SDK_ID      = "SDK_ID_GO_PROVIDER"
```
Comment (nicklasl) [inline confidence/confidence.go:22]: Check this ID. Is there another one to set for Confidence SDK?

File: confidence/confidence.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+func (e Confidence) PutContext(key string, value interface{}) {
+	e.contextMap[key] = value
+}
+
+func (e Confidence) Track(ctx context.Context, eventName string, message map[string]interface{}) {
+	go func() {
+		currentTime := time.Now()
+		iso8601Time := currentTime.Format(time.RFC3339)
+
+		newMap := e.GetContext()
```
Comment (nicklasl) [inline confidence/confidence.go:83]: I suggest we move the `getContext` call outside of the go routine.
Maybe best to actually only wrap the long running upload in it?

File: confidence/confidence_internal_test.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	MockedError    error
+	TestingT       *testing.T
+}
+
+func (r MockResolveClient) SendResolveRequest(_ context.Context,
+	request ResolveRequest) (ResolveResponse, error) {
+	assert.Equal(r.TestingT, "user1", request.EvaluationContext["targeting_key"])
+	return r.MockedResponse, r.MockedError
+}
+
```
Comment (nicklasl) [inline confidence/confidence_internal_test.go:24]: Would be super happy if we had some tests for the parent -> child context inheritance and shadowing stuff. 🙏

Comment (vahidlazio) [reply confidence/confidence_internal_test.go:24]: will add some.

File: confidence/confidence.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+func NewConfidenceBuilder() ConfidenceBuilder {
+	return ConfidenceBuilder{
+		confidence: Confidence{},
+	}
+}
+
+func (e Confidence) PutContext(key string, value interface{}) {
+	e.contextMap[key] = value
```
Comment (nicklasl) [inline confidence/confidence.go:75]: Does this function support "shadowing" a parent context key with a null object?

Comment (vahidlazio) [reply confidence/confidence.go:75]: this just replaces whatever the user passes, then when reading this context from this instance, the context from this instance will take priority over the one coming from the backend. was this what you asked?

File: confidence/confidence.go (hunk 1 above)
Comment (vahidlazio) [reply confidence/confidence.go:22]: i think it's _Confidence instead of provider.

File: confidence/confidence_internal_test.go (hunk 3 above)
Comment (vahidlazio) [reply confidence/confidence_internal_test.go:24]: done.

File: confidence/confidence.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+	ResolveClient ResolveClient
+}
+
+func (e Confidence) GetContext() map[string]interface{} {
+	currentMap := map[string]interface{}{}
+	parentMap := make(map[string]interface{})
+	if e.parent != nil {
+		parentMap = e.parent.GetContext()
+	}
+for key, value := range parentMap {
```
Comment (nicklasl) [inline confidence/confidence.go:41]: nit: indentation 🫨 

is there something we can add to the CI to check these things?

Comment (vahidlazio) [reply confidence/confidence.go:41]: it should be some command as well to do it for the whole project. let me dig a bit.

Comment (vahidlazio) [reply confidence/confidence.go]: done using `gofmt -w`, also added a step in the ci to check the formatting and fail the step if formatting is not correct.

**Review by nicklasl — APPROVED** [review]

**Pull request merged**